sklearn/metrics/_scorer.py

```diff
@@ -711,40 +639,65 @@ def make_scorer(
-    output of :term:`decision_function` or :term:`predict_proba` when
-    :term:`decision_function` is not present.
-
     Examples
     --------
     >>> from sklearn.metrics import fbeta_score, make_scorer
     >>> ftwo_scorer = make_scorer(fbeta_score, beta=2)
     >>> ftwo_scorer
-    make_scorer(fbeta_score, beta=2)
+    make_scorer(fbeta_score, response_method='predict', beta=2)
     >>> from sklearn.model_selection import GridSearchCV
     >>> from sklearn.svm import LinearSVC
     >>> grid = GridSearchCV(LinearSVC(), param_grid={'C': [1, 10]},
     ...                     scoring=ftwo_scorer)
     """
-    sign = 1 if greater_is_better else -1
-    if needs_proba and needs_threshold:
+    if response_method is not None and (
+        needs_proba != "deprecated" or needs_threshold != "deprecated"
+    ):
         raise ValueError(
-            "Set either needs_proba or needs_threshold to True, but not both."
+            "You cannot set both `response_method` and `needs_proba` or "
+            "`needs_threshold` at the same time. Only use `response_method` since "
+            "the other two are deprecated in version 1.4 and will be removed in 1.6."
         )
-    if needs_proba:
-        cls = _ProbaScorer
-    elif needs_threshold:
-        cls = _ThresholdScorer
-    else:
-        cls = _PredictScorer
-    return cls(score_func, sign, kwargs)
+    if needs_proba is True and needs_threshold is True:
+        raise ValueError(
+            "You cannot set both `needs_proba` and `needs_threshold` at the same "
+            "time. Use `response_method` instead since the other two are deprecated "
+            "in version 1.4 and will be removed in 1.6."
+        )
+
+    if needs_proba is False and needs_threshold is False:
+        warnings.warn(
+            (
+                "The `needs_threshold` and `needs_proba` parameter are deprecated in "
+                "version 1.4 and will be removed in 1.6. You can either let "
+                "`response_method` be `None` or set it to `predict` to preserve the "
+                "same behaviour."
+            ),
+            FutureWarning,
+        )
+        response_method = "predict"
+    elif needs_proba != "deprecated":
+        warnings.warn(
+            (
+                "The `needs_proba` parameter is deprecated in version 1.4 and will be "
+                'removed in 1.6. Use `response_method="predict_proba"` instead.'
+            ),
+            FutureWarning,
+        )
+        response_method = "predict_proba"
+    elif needs_threshold != "deprecated":
+        warnings.warn(
+            (
+                "The `needs_threshold` parameter is deprecated in version 1.4 and will"
+                ' be removed in 1.6. Use `response_method="decision_function"` instead.'
+            ),
+            FutureWarning,
+        )
+        response_method = ("decision_function", "predict_proba")
+
+    response_method = "predict" if response_method is None else response_method
```

Comment:
The implementation here is confusing and buggy. For instance, check when: `response_method=None, needs_proba=False, needs_threshold=True`. Then response method would be  `"predict_proba"`.

I suggest a function like this for this purpose, to encapsulate it, and I find it a bit more clear too:

```py
def get_response_method(response_method, needs_threshold, needs_proba):
    needs_threshold_provided = needs_threshold != "deprecated"
    needs_proba_provided = needs_proba != "deprecated"
    response_method_provided = response_method is not None

    needs_threshold = False if needs_threshold == "deprecated" else needs_threshold
    needs_proba = False if needs_proba == "deprecated" else needs_proba

    if response_method_provided and (needs_proba_provided or needs_threshold_provided):
        raise ValueError(
            "You cannot set both `response_method` and `needs_proba` or "
            "`needs_threshold` at the same time. Only use `response_method` since "
            "the other two are deprecated in version 1.4 and will be removed in 1.6."
        )

    if needs_proba_provided or needs_threshold_provided:
        warnings.warn(
            (
                "The `needs_threshold` and `needs_proba` parameter are deprecated in "
                "version 1.4 and will be removed in 1.6. You can either let "
                "`response_method` be `None` or set it to `predict` to preserve the "
                "same behaviour."
            ),
            FutureWarning,
        )

    if response_method_provided:
        return response_method

    if needs_proba is True and needs_threshold is True:
        raise ValueError(
            "You cannot set both `needs_proba` and `needs_threshold` at the same "
            "time. Use `response_method` instead since the other two are deprecated "
            "in version 1.4 and will be removed in 1.6."
        )

    if needs_proba is True:
        response_method = "predict_proba"
    elif needs_threshold is True:
        response_method = ("decision_function", "predict_proba")
    else:
        response_method = "predict"

    return response_method
```

Reply:
Yep, I added new test to cover the missing cases.